File: experiments/Layout.hs
module Layout where

import Data.Ratio
import Data.Maybe
import Data.Array
import Control.Arrow (first)
import Control.Monad (mapM_)

-- How much relative space a plot takes up in the layout direction.
type Space = Maybe (Ratio Int)

-- A layout is made up of frames arranged in horizontal boxes and
-- vertical boxes.  Frames are just their titles here.
data Layout a = HBox [(a, Layout a)]
              | VBox [(a, Layout a)]
              | Frame String
              deriving (Eq, Show)

-- Specialisation for the input type of layouts where we specify
-- ratios of the available space used by frames, and the final layed
-- out version, which gives the exact sizes of everything.
type LayoutIn = Layout Space
type Layedout = Layout Int

-- Fix ratios...
fixrat :: Ratio Int -> [(Space, LayoutIn)] -> [(Space, LayoutIn)]
fixrat r = map (first (fmap (* r)))

-- Operators for laying out frames.
(|||) :: LayoutIn -> LayoutIn -> LayoutIn
(HBox ps)   ||| p@(Frame _) = HBox $ ps ++ [(Nothing, p)]
p@(Frame _) ||| (HBox ps)   = HBox $ (Nothing, p) : ps
p1          ||| p2          = HBox [(Nothing, p1), (Nothing, p2)]

(%||) :: (Ratio Int, LayoutIn) -> LayoutIn -> LayoutIn
(r, HBox ps)     %|| p@(Frame _) = HBox $ fixrat r ps ++ [(Nothing, p)]
(r, p@(Frame _)) %|| (HBox ps)   = HBox $ (Just r, p) : ps
(r, p1)          %|| p2          = HBox [(Just r, p1), (Nothing, p2)]

(%|%) :: (Ratio Int, LayoutIn) -> (Ratio Int, LayoutIn) -> LayoutIn
(r1, HBox ps)     %|% (r2, p@(Frame _)) = HBox $ fixrat r1 ps ++ [(Just r2, p)]
(r1, p@(Frame _)) %|% (r2, HBox ps)     = HBox $ (Just r1, p) : fixrat r2 ps
(r1, p1)          %|% (r2, p2)          = HBox [(Just r1, p1), (Just r2, p2)]

(||%) :: LayoutIn -> (Ratio Int, LayoutIn) -> LayoutIn
p@(Frame _) ||% (r, HBox ps)     = HBox $ (Nothing, p) : fixrat r ps
(HBox ps)   ||% (r, p@(Frame _)) = HBox $ ps ++ [(Just r, p)]
p2          ||% (r, p1)          = HBox [(Nothing, p1), (Just r, p2)]

(===) :: LayoutIn -> LayoutIn -> LayoutIn
(VBox ps)   === p@(Frame _) = VBox $ ps ++ [(Nothing, p)]
p@(Frame _) === (VBox ps)   = VBox $ (Nothing, p) : ps
p1          === p2          = VBox [(Nothing, p1), (Nothing, p2)]

(%==) :: (Ratio Int, LayoutIn) -> LayoutIn -> LayoutIn
(r, VBox ps)     %== p@(Frame _) = VBox $ fixrat r ps ++ [(Nothing, p)]
(r, p@(Frame _)) %== (VBox ps)   = VBox $ (Just r, p) : ps
(r, p1)          %== p2          = VBox [(Just r, p1), (Nothing, p2)]

(%=%) :: (Ratio Int, LayoutIn) -> (Ratio Int, LayoutIn) -> LayoutIn
(r1, VBox ps)     %=% (r2, p@(Frame _)) = VBox $ fixrat r1 ps ++ [(Just r2, p)]
(r1, p@(Frame _)) %=% (r2, VBox ps)     = VBox $ (Just r1, p) : fixrat r2 ps
(r1, p1)          %=% (r2, p2)          = VBox [(Just r1, p1), (Just r2, p2)]

(==%) :: LayoutIn -> (Ratio Int, LayoutIn) -> LayoutIn
(VBox ps)   ==% (r, p@(Frame _)) = VBox $ ps ++ [(Just r, p)]
p@(Frame _) ==% (r, VBox ps)     = VBox $ (Nothing, p) : fixrat r ps
p1          ==% (r, p2)          = VBox [(Nothing, p1), (Just r, p2)]


-- Some frames.
a, b, c, d :: LayoutIn
a = Frame "A"
b = Frame "B"
c = Frame "C"
d = Frame "D"

-- Some layouts.
l1, l2, l3, l4, l5, l6 :: LayoutIn
l1 = a ||| b
l2 = a === b
l3 = (a ||| b) === (c ||| d)
l4 = (a === b === c) ||| d
l5 = (3, a) %|% (2, b)
l6 = a ==% (1,b) ==% (3,c)

-- Determine sizes of frames in a layout.
calcSizes :: (Int, Int, Int) -> LayoutIn -> Layedout
calcSizes (w, h, spc) (Frame s) = HBox [(w, VBox [(h, Frame s)])]
calcSizes (w, h, spc) p@(HBox _) = VBox [(h, calcSizes' (w, h, spc) p)]
calcSizes (w, h, spc) p@(VBox _) = HBox [(w, calcSizes' (w, h, spc) p)]

calcSizes' :: (Int, Int, Int) -> LayoutIn -> Layedout
calcSizes' (w, h, spc) (HBox ps) = HBox $ zipWith recurse sizes (map snd ps)
  where sizes = fitSizes w spc $ map fst ps
        recurse fw p = (fw, calcSizes' (fw, h, spc) p)
calcSizes' (w, h, spc) (VBox ps) = VBox $ zipWith recurse sizes (map snd ps)
  where sizes = fitSizes h spc $ map fst ps
        recurse fh p = (fh, calcSizes' (w, fh, spc) p)
calcSizes' (_, _, _) (Frame s) = Frame s

-- Fit a range of space parameters into a given size and spacing.
fitSizes :: Int -> Int -> [Space] -> [Int]
fitSizes w s ws = widths
  where realw = w - s * (length ws - 1)
        mean = case catMaybes ws of
          [] -> 1
          vs -> sum vs / (fromIntegral $ length vs)
        realvals = map (maybe mean id) ws
        realtot = sum realvals
        tmpwidths = map (floor . (* (realw % 1)) . (/ realtot)) realvals
        wdiff = realw - sum tmpwidths
        spc = length ws `div` wdiff
        ds = cycle (1 : take spc (repeat 0))
        widths = if wdiff == 0 then tmpwidths else zipWith (+) tmpwidths ds

-- Render a layout as strings.
doRender :: (Int,Int,Int) -> Layedout -> [String]
doRender (w,h,s) l = toStrings (w,h) $ foldl renderOne empty frames
  where empty = listArray ((1,1), (w,h)) (repeat ' ')
        frames = extractFrames s (1,1) (w,h) l

-- Full render.
render :: (Int,Int,Int) -> LayoutIn -> IO ()
render (w,h,s) ls = mapM_ putStrLn $ doRender (w,h,s) $ calcSizes (w,h,s) ls

type View = (Int,Int,Int,Int,String)
type Board = Array (Int,Int) Char

-- Extract frames with positions from layout.
extractFrames :: Int -> (Int,Int) -> (Int,Int) -> Layedout -> [View]
extractFrames spc (curx,cury) (_,h) (HBox ps) =
  snd $ foldl addone ((curx,cury),[]) ps
  where addone ((x,y),acc) (fw,Frame s) = ((x+fw+spc,y), acc ++ [(x,y,fw,h,s)])
        addone ((x,y),acc) (vw,vb@(VBox _)) =
          ((x+vw+spc,y), acc ++ extractFrames spc (x,y) (vw,h) vb)
extractFrames spc (curx,cury) (w,_) (VBox ps) =
  snd $ foldl addone ((curx,cury),[]) ps
  where addone ((x,y),acc) (fh,Frame s) = ((x,y+fh+spc), acc ++ [(x,y,w,fh,s)])
        addone ((x,y),acc) (hh,hb@(HBox _)) =
          ((x,y+hh+spc), acc ++ extractFrames spc (x,y) (w,hh) hb)

-- Render one frame.
renderOne :: Board -> View -> Board
renderOne arr (x,y,w,h,s) = arr // (corners ++ sides ++ label)
  where corners = zip [(x,y),(x+w-1,y),(x,y+h-1),(x+w-1,y+h-1)] (repeat '+')
        sides = lside ++ rside ++ tside ++ bside
        lside = [((x,iy),'|')     | iy <- [y+1..y+h-2]]
        rside = [((x+w-1,iy),'|') | iy <- [y+1..y+h-2]]
        tside = [((ix,y),'-')     | ix <- [x+1..x+w-2]]
        bside = [((ix,y+h-1),'-') | ix <- [x+1..x+w-2]]
        label = zip lcoords s
        lcoords = [(ix,y+h`div`2) | ix <- [x+w`div`2-(length s)`div`2..]]

-- Convert array representation to strings.
toStrings :: (Int,Int) -> Board -> [String]
toStrings (w,h) arr = [[ arr!(col,row) | col <- [1..w]] | row <- [1..h]]

File: src/layout.js
// Deal with plot layout and processing for <plot-row>, <plot-col>,
// <plot-grid> and <plot-layout> directives.


// How much relative space a plot takes up in the current layout
// direction is specified as a number or nothing (if the plot will be
// sized according to the average size of plots in the box).
//
// A layout is made up of plots arranged in horizontal boxes and
// vertical boxes.  A single <plot> directive manages its own layout
// (which is trivial).
//
// Within the data structures for managing layouts, individual plots
// are represented by their associated Angular scopes.
//
// If we have some plots, A, B, C, D, which we'll represent by
// $scopeA, $scopeB, etc., then here's how we represent some example
// layouts:
//
//  +---+ +---+
//  | A | | B |    ["hbox", [$scopeA, $scopeB]]
//  +---+ +---+
//
//  +---+
//  | A |          ["vbox", [$scopeA, $scopeB]]
//  +---+
//  +---+
//  | B |
//  +---+
//
//  +---+ +---+
//  | A | | B |    ["vbox", ["hbox", [$scopeA, $scopeB]],
//  +---+ +---+             ["hbox", [$scopeC, $scopeD]]]
//  +---+ +---+
//  | C | | D |
//  +---+ +---+
//
//  +---+ +---+
//  | A | |   |    ["hbox", ["vbox", [$scopeA, $scopeB, $scopeC]],
//  +---+ |   |             $scopeD]
//  +---+ |   |
//  | B | | D |
//  +---+ |   |
//  +---+ |   |
//  | C | |   |
//  +---+ +---+
//
//  +-------+ +----+
//  |       | |    |    ["hbox", [[3, $scopeA], [2, $scopeB]]]
//  |   A   | |  B |
//  |       | |    |
//  +-------+ +----+
//
//  +--------------+    (Here, B is supposed to be a box two rows high.)
//  |      A       |
//  |              |    ["vbox", [[2, $scopeA], [1, $scopeB], [3, $scopeC]]]
//  +--------------+
//  +------B-------+
//  +--------------+
//  +--------------+
//  |              |
//  |              |
//  |      C       |
//  |              |
//  |              |
//  +--------------+
//
// In Haskell, a layout is defined as
//
//   data Layout a = HBox [(a, Layout a)]
//                 | VBox [(a, Layout a)]
//                 | Frame String
//                 deriving (Eq, Show)
//
// where a is a type parameterising the size specifications for the
// individual plots.  As specified by the user, a ~ Maybe (Ratio Int),
// and using Nothing means that a plot should just take up the
// "average" space.  For rendering, the Layout (Maybe (Ratio Int))
// type is transformed into a Layout Int which specifies the actual
// dimensions of the plots.  This transformation is the job of the
// layoutSizes function.


radian.factory('layoutSizes', ['layoutToString', function(layoutToString) {
  'use strict';

  // Determine sizes of frames in a layout.
  return function(w, h, spc, layout)
  {
    // Fit a range of space parameters into a given size and spacing.
    function fitSizes(w, ws)
    {
      var realw = w - spc * (ws.length - 1);
      var vs = ws.filter(function(x) { return x != null; });
      function sum(a) { return a.reduce(function(a,b) { return a+b; }, 0); };
      var mean = vs.length == 0 ? 1 : sum(vs) / vs.length;
      var realvals = ws.map(function(x) { return x == null ? mean : x; });
      var realtot = sum(realvals);
      var widths = realvals.map(function(x) {
        return Math.floor(x * realw / realtot);
      });
      var wdiff = realw - sum(widths);
      if (wdiff != 0) {
        var outspc = Math.floor(ws.length / wdiff);
        for (var i = 0; i < ws.length; ++i)
          if (i % (outspc + 1) == 0) ++widths[i];
      }
      return widths;
    };

    function help(w, h, layout)
    {
      if (layout.type == 'plot' || layout.type == 'empty') return layout;
      function getratios(ls) {
        return ls.map(function(l) { return l.size || null; });
      };
      var sizes =
        fitSizes(layout.type == 'hbox' ? w : h, getratios(layout.items));
      var outitems = [];
      for (var i = 0; i < layout.items.length; ++i) {
        if (layout.type == 'hbox')
          outitems[i] = { size: sizes[i],
                          item: help(sizes[i], h, layout.items[i].item) };
        else
          outitems[i] = { size: sizes[i],
                          item: help(w, sizes[i], layout.items[i].item) };
      }
      return { type: layout.type, items: outitems };
    };

    var sub = help(w, h, layout);
    if (layout.type == 'hbox')
      return { type: 'vbox', items: [{ size: h, item: sub }] };
    else if (layout.type == 'vbox')
      return { type: 'hbox', items: [{ size: w, item: sub }] };
    else
      return { type:'hbox',
               items: [{ size: w, item: { type: 'vbox', items: sub }}]};
  };
}]);

radian.factory('addToLayout', function()
{
  return function(sc, sublayout, size) {
    if (sublayout.hasOwnProperty('$id'))
      sc.layoutItems.push({ size: size != null ? Number(size) : null,
                            item: { type: 'plot', items: sublayout } });
    else
      sc.layoutItems.push({ size: size != null ? Number(size) : null,
                            item: sublayout });
  };
});

radian.factory('extractFrames',
 ['layoutToString',
  function(layoutToString)
{
  // A "frame" is an object of the form { x, y, w, h, plot }, where
  // plot points to the plot scope.
  return function(spc, w, h, layout) {
    function go(curx, cury, curw, curh, lay) {
      var frames = [];
      if (lay.type == 'hbox') {
        for (var i = 0; i < lay.items.length; ++i) {
          var item = lay.items[i].item;
          var itype = lay.items[i].item.type;
          var isize = lay.items[i].size;
          if (item.type == 'plot') {
            frames.push({ x: curx, y: cury, w: isize, h: curh,
                          plot: item.items });
          } else if (item.type == 'vbox') {
            frames = frames.concat(go(curx, cury, isize, curh, item));
          }
          curx += isize + spc;
        }
      } else if (lay.type == 'vbox') {
        for (var i = 0; i < lay.items.length; ++i) {
          var item = lay.items[i].item;
          var itype = lay.items[i].item.type;
          var isize = lay.items[i].size;
          if (item.type == 'plot') {
            frames.push({ x: curx, y: cury, w: curw, h: isize,
                          plot: item.items });
          } else if (item.type == 'hbox') {
            frames = frames.concat(go(curx, cury, curw, isize, item));
          }
          cury += isize + spc;
        }
      } else throw Error("invalid layout passed to extractFrames");
      return frames;
    };
    return go(0, 0, w, h, layout);
  };
}]);

radian.factory('layoutDirective',
 ['layoutSizes', 'processAttrs', 'calcPlotDimensions',
  'addToLayout', 'extractFrames', 'layoutToString',
  function(layoutSizes, processAttrs, calcPlotDimensions,
           addToLayout, extractFrames, layoutToString)
{
  'use strict';

  return function(container) {
    function preLink(sc, elm, as, transclude) {
      processAttrs(sc, as);
      if (!sc.inLayout) {
        sc.layoutTop = true;
        sc.inLayout = true;
        if (!sc.inStack) calcPlotDimensions(sc, elm, as);
        $(elm).css('width', sc.width).css('height', sc.height);
        sc.layoutsvg = elm.children()[0];
      } else
        $(elm.children()[1]).remove();
      sc.layoutItems = [];
      transclude(sc.$new(), function (cl) { elm.append(cl); });
    };

    function postLink(sc, elm) {
      var items = { type: container, items: sc.layoutItems };
      if (sc.hasOwnProperty('layoutTop')) {
        var spacing = sc.spacing || 0;
        var layedout = layoutSizes(sc.width, sc.height, spacing, items);
        var frames = extractFrames(0, sc.width, sc.height, layedout);
        if (sc.hasOwnProperty('title')) items.title = sc.title;
        frames.forEach(function(fr) {
          fr.plot.width = fr.w;
          fr.plot.height = fr.h;
          fr.plot.svg = d3.select(sc.layoutsvg).append('g')
            .attr('width', fr.w).attr('height', fr.h)
            .attr('transform', 'translate(' + fr.x + ',' + fr.y + ')')[0][0];
        });
      }
      if (!sc.hasOwnProperty('layoutTop') || sc.inStack)
        addToLayout(sc.$parent, items, sc.layoutShare);
    };

    return {
      restrict: 'E',
      template: '<div class="radian"><svg></svg></div>',
      replace: true,
      transclude: true,
      scope: true,
      compile: function(elm, as, trans) {
        return { pre: function(s, e, a) { preLink(s, e, a, trans); },
                 post: postLink };
      }
    };
  };
}]);

radian.directive('plotRow', ['layoutDirective', function(layoutDirective)
{
  return layoutDirective('hbox');
}]);

radian.directive('plotCol', ['layoutDirective', function(layoutDirective)
{
  return layoutDirective('vbox');
}]);

radian.directive('plotGrid',
 ['layoutSizes', 'processAttrs', 'calcPlotDimensions',
  'addToLayout', 'extractFrames', 'layoutToString',
  function(layoutSizes, processAttrs, calcPlotDimensions,
           addToLayout, extractFrames, layoutToString)
{
  'use strict';

  function preLink(sc, elm, as, transclude) {
    processAttrs(sc, as);
    if (!sc.inLayout) {
      sc.layoutTop = true;
      sc.inLayout = true;
      if (!sc.inStack) calcPlotDimensions(sc, elm, as);
      $(elm).css('width', sc.width).css('height', sc.height);
      sc.layoutsvg = elm.children()[0];
    } else
      $(elm.children()[1]).remove();
    sc.layoutItems = [];
    transclude(sc.$new(), function (cl) { elm.append(cl); });
  };

  function postLink(sc, elm) {
    var nrows = sc.rows || Math.floor(Math.sqrt(sc.layoutItems.length));
    var ncols = sc.cols || Math.ceil(sc.layoutItems.length / nrows);
    var rows = [];
    var i = 0;
    for (var r = 0; r < nrows; ++r) {
      var cols = [];
      for (var c = 0; c < ncols; ++c) {
        if (i >= sc.layoutItems.length)
          cols.push({ size: null, item: { type: 'empty' } });
        else
          cols.push(sc.layoutItems[i++]);
      }
      rows.push({ size: null, item: { type: 'hbox', items: cols } });
    }
    var items = { type: 'vbox', items: rows };
    if (sc.hasOwnProperty('layoutTop')) {
      var spacing = sc.spacing || 0;
      var layedout = layoutSizes(sc.width, sc.height, spacing, items);
      var frames = extractFrames(0, sc.width, sc.height, layedout);
      if (sc.hasOwnProperty('title')) items.title = sc.title;
      frames.forEach(function(fr) {
        fr.plot.width = fr.w;
        fr.plot.height = fr.h;
        fr.plot.svg = d3.select(sc.layoutsvg).append('g')
          .attr('width', fr.w).attr('height', fr.h)
          .attr('transform', 'translate(' + fr.x + ',' + fr.y + ')')[0][0];
      });
    }
    if (!sc.hasOwnProperty('layoutTop') || sc.inStack)
      addToLayout(sc.$parent, items, sc.layoutShare);
  };

  return {
    restrict: 'E',
    template: '<div class="radian"><svg></svg></div>',
    replace: true,
    transclude: true,
    scope: true,
    compile: function(elm, as, trans) {
      return { pre: function(s, e, a) { preLink(s, e, a, trans); },
               post: postLink };
    }
  };
}]);


radian.directive('plotStack',
 ['$rootScope', 'layoutSizes', 'processAttrs', 'calcPlotDimensions',
  'addToLayout', 'extractFrames', 'layoutToString',
  function($rootScope, layoutSizes, processAttrs, calcPlotDimensions,
           addToLayout, extractFrames, layoutToString)
{
  'use strict';

  function preLink(sc, elm, as, transclude) {
    processAttrs(sc, as);
    if (sc.inLayout)
      throw Error("<plot-stack> cannot appear inside other layout directives");
    if (!sc.inStack) calcPlotDimensions(sc, elm, as);
    if (sc.inStack) addToLayout(sc.$parent, { type: 'stack', items: sc }, null);
    sc.inStack = true;
    sc.layoutItems = [];
    if (!$rootScope.radianNavIDs) $rootScope.radianNavIDs = { };
    transclude(sc.$new(), function (cl) {
      elm.append('<div class="tab-content radian-tabs"></div>');
      var tabs = elm.children(0);
      tabs.append(cl);
      sc.ids = [];
      cl.filter('div.radian,div.radian-stack').each(function(i) {
        var idx = 0, tabid;
        do {
          ++idx;
          tabid = 'tab' + idx + '_' + i;
        } while ($rootScope.radianNavIDs[tabid]);
        sc.ids.push(tabid);
        $rootScope.radianNavIDs[tabid] = 1;
        var cls = i == 0 ? 'tab-pane active' : 'tab-pane';
        $(this).wrap('<div class="' + cls + '" id="' + tabid + '"></div>');
      });
    });
  };

  function postLink(sc, elm) {
    var is = sc.layoutItems;
    elm.prepend('<ul class="nav nav-tabs"></ul>');
    var nav = elm.children('ul');
    for (var i = 0; i < is.length; ++i) {
      var it = is[i].item;
      var t = it.title ? it.title :
        (it.items.title ? it.items.title : 'Tab ' + (i+1));
      var link = '<a href="#' + sc.ids[i] + '" data-toggle="tab">' + t + '</a>';
      var active = i == 0 ? ' class="active"' : '';
      nav.append('<li' + active + '>' + link + '</li>');
    }
  };

  return {
    restrict: 'E',
    template: '<div class="radian-stack"></div>',
    replace: true,
    transclude: true,
    scope: true,
    compile: function(elm, as, trans) {
      return { pre: function(s, e, a) { preLink(s, e, a, trans); },
               post: postLink };
    }
  };
}]);




radian.factory('layoutToString', function() {
  'use strict';

  return function(layout) {
    function fixplots(lay) {
      switch (lay.type) {
      case 'empty': return { type: 'empty' };
      case 'plot': return { type: 'plot', items: lay.items.$id };
      default: return { type: lay.type, items: lay.items.map(function(i) {
        return { size: i.size, item: fixplots(i.item) };
      }) };
      }
    };
    return JSON.stringify(fixplots(layout));
  };
});
